grpc/dispatcher.go

```diff
@@ -61,9 +57,20 @@ func (d *Dispatcher) Do(request fiber.Request) fiber.Response {
 	defer cancel()
 	ctx = metadata.NewOutgoingContext(ctx, grpcRequest.Metadata)
 
-	responseProto := proto.Clone(d.responseProto)
+	response := new(bytes.Buffer)
 	var responseHeader metadata.MD
-	err := d.conn.Invoke(ctx, d.serviceMethod, grpcRequest.Payload(), responseProto, grpc.Header(&responseHeader))
+
+	// Dispatcher will send both request and payload as bytes, with the use of codec
+	// to prevent marshaling. The codec content type will be sent with request and
+	// the server will attempt to unmarshal with the codec.
+	err := d.conn.Invoke(
+		ctx,
+		d.serviceMethod,
+		grpcRequest.Payload(),
+		response,
+		grpc.Header(&responseHeader),
+		grpc.CallContentSubtype(CodecName),
```

Comment:
thats an interesting catch. Within Turing, it will receive `application/grpc+fibercodec` from Fiber.
![Screenshot 2022-09-16 at 10 14 35 AM](https://user-images.githubusercontent.com/30390872/190541871-645827ea-7724-4d19-bc81-f38c0d05c61c.png)

However since the client calling Turing did not specific a codec, Turing doesn't reply with a codec, so `fibercodec` stays between Turing and Fiber, and the content type is overwritten
<img width="1122" alt="Screenshot 2022-09-16 at 10 14 49 AM" src="https://user-images.githubusercontent.com/30390872/190542022-7edc143f-db56-4992-ba19-a04dacee5e56.png">

Reply:
Will routes/model service receive `application/grpc+fibercodec` content type header? I think we should avoid that since it might confuse the receiving end, some grpc server implementation might not work well and return error.

Reply:
yes they will. The only way to hack around this is to replace the default codec with fiber implementation which can introduce more unintended side effects since it impact the entire application, if the default codec is expected to be used in other area. 

As per grpc library [guideline](https://github.com/grpc/grpc-go/blob/master/Documentation/encoding.md)
`Codecs, by definition, must be symmetric, so the same desired Codec should be registered in both client and server binaries.` If we are using another codec, we ought to let downstream be aware and for them to handle it. 

Grpc server implementation will always call [getCodec](https://github.com/grpc/grpc-go/blob/master/server.go#L1295) which will never return nil. In this case, it will be called with `fibercodec` and if it is not registered in the downstream, will return `return encoding.GetCodec(proto.Name)` which is the default codec.

```
// contentSubtype must be lowercase
// cannot return nil
func (s *Server) getCodec(contentSubtype string) baseCodec {
	if s.opts.codec != nil {
		return s.opts.codec
	}
	if contentSubtype == "" {
		return encoding.GetCodec(proto.Name)
	}
	codec := encoding.GetCodec(contentSubtype)
	if codec == nil {
		return encoding.GetCodec(proto.Name)
	}
	return codec
}
```

Reply:
> Grpc server implementation will always call [getCodec](https://github.com/grpc/grpc-go/blob/master/server.go#L1295) which will never return nil. In this case, it will be called with fibercodec and if it is not registered in the downstream, will return return encoding.GetCodec(proto.Name) which is the default codec.

Can we confirm this for other languages?

Reply:
trying to dig python grpc code for long time and there doesn't seems to be `codec` available there. 

My best guess is the [rpc.unary_unary_rpc_method_handler](https://grpc.github.io/grpc/python/grpc.html#grpc.unary_unary_rpc_method_handler) which accept serialiser. 
It only does string serialisation in the [generated proto](https://github.com/caraml-dev/universal-prediction-interface/blob/main/gen/python/grpc/caraml/upi/v1/upi_pb2_grpc.py#L40)

I tested with the generated grpc in both go and python, sending a request with metadata `model=abc`

For python server using 
```
        for key, value in context.invocation_metadata():
            print('Received initial metadata: key=%s value=%s' % (key, value))
```
<img width="615" alt="image" src="https://user-images.githubusercontent.com/30390872/190583450-14c6f4de-3eb5-42d1-97cc-5dcb2e066266.png">


For go server using
```
metadata.FromIncomingContext(ctx)
```
```
accept-encoding:[identity] 
content-type:[application/grpc+fibercodec] 
grpc-accept-encoding:[identity,deflate,gzip] 
model:[abc] 
turing-req-id:[6a3ed539-7872-4c70-bd4d-d5c3b6a0ba74] 
user-agent:[grpc-go/1.48.0]]
```
![image](https://user-images.githubusercontent.com/30390872/190583417-fb8a3bea-cd70-4beb-90fd-8b21a376e585.png)

There some slight different, golang headers has `content-type` and more `grpc-accept-encoding` values. The rest are the same (both receive the model metadata sent from initial client and turing-req-id).

Interestingly for Java, there is no `contentSubType` support and there a [ticket for it](https://github.com/grpc/grpc-java/issues/7321)

There are some interesting comments,

```
Java has Codec, CompressorRegistry and DecompressorRegistry which doesn't use contentSubType 
but rather the "grpc-encoding" header and is mainly used for compression/decompression rather 
than custom ser-deser. Would that work for you?
```
```
grpc-java does not support the subtype thing; it is currently ignored and never sent. 
grpc-go added support for it thinking other languages supported it, when they don't. 
It would be possible to add, but would require additions to how MethodDescriptor.Marshallers 
work and would be quite different to set up than Go's RegisterCodec as marshallers are
not global in Java 
```

Just judging from java and python, i think this `codec/contentSubType` is used exclusively by golang and is not support and ignored in other language (like the python server example)

Reply:
Thanks for confirming. I am still not fully convinced it will be safe though, but I guess we can go ahead.